Dolphin scripts/Entrance Randomizer/lib/utils.py

```diff
@@ -92,17 +92,88 @@ def follow_pointer_path(ppath: Sequence[int]):
     return addr
 
 
+def prevent_item_softlock():
+    """
+    Prevent softlocking by missing the right items.
+
+    Even with logic this can happen if the player bypasses logic.
+    """
+    # TODO: A better fix would be to just teleport the player to the entrance of the temple,
+    # but we can't reliably do that yet.
+    # And we can't use position-based fix as the game checks for walls even when teleporting
+
+    # Pickaxe lets you escape everything
+    if memory.read_u32(ADDRESSES.backpack_struct + BackpackOffset.Pickaxes):
+        return
+
+    # Scorpion Temple
+    if (
+        not memory.read_u32(ADDRESSES.backpack_struct + BackpackOffset.Torch)
+        and state.current_area_old == LevelCRC.SCORPION_SPIRIT
+        and state.current_area_new == LevelCRC.SCORPION_TEMPLE
+    ):
+        # Lets just give the player the torch, it's not like it unlocks much outside convenience.
+        memory.write_u32(ADDRESSES.backpack_struct + BackpackOffset.Torch, 1)
+        return
+
+    # Penguin Temple: https://youtu.be/LHglikRqeAw
+
+    # Apu Illapu Shrine
+    if (
+        state.area_load_state_new == 6  # noqa: PLR2004
+        and state.current_area_new == LevelCRC.APU_ILLAPU_SHRINE
+        and not memory.read_u32(ADDRESSES.backpack_struct + BackpackOffset.TNT)
+        and not memory.read_u32(
+            follow_pointer_path((ADDRESSES.player_ptr, PlayerPtrOffset.Breakdance)),
+        )
+        # Item sliding check
+        and not (
+            memory.read_u32(
+                follow_pointer_path((ADDRESSES.player_ptr, PlayerPtrOffset.HeroicDash)),
+            ) and any(
+                memory.read_u32(ADDRESSES.backpack_struct + item)
+                for item in (
+                    BackpackOffset.Canteen,
+                    BackpackOffset.Slingshot,
+                    BackpackOffset.Torch,
+                    BackpackOffset.Shield,
+                    BackpackOffset.GasMask,
+                    BackpackOffset.TNT,
+                )
+            )
+        )
+    ):
```

Comment:
You can also beat the spinjas if you have Sling + Super Sling, so no need to kick you out in that situation.

Reply:
Yes killing all Spinjas with Super Sling only is realistic, as I've shown a year ago: https://youtu.be/veWV_96UdIg?si=T1jcFbgS7zi-sDlN&t=1732
And even though it **feels** painfully slow because I miss 90% of my shots, I still beat all 10 Spinjas in only 3 minutes, which is a whole lot faster than what we have to deal with Shield-only!

Not including Shield TNT bounce makes perfect sense, because that's still a random factor depending on Mole's throwing directions. Sometimes entering that level with nothing but Shield is perfectly fine, but other times it's a softlock. So I would agree with you to not consider the Shield enough to enter Apu Illapu Shrine normally.